src/app/ReadClient.h

```diff
@@ -0,0 +1,207 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines read client for a CHIP Interaction Data model
+ *
+ */
+
+#pragma once
+
+#ifndef _CHIP_INTERACTION_MODEL_READ_CLIENT_H
+#define _CHIP_INTERACTION_MODEL_READ_CLIENT_H
+
+#include <app/MessageDef/ReadRequest.h>
+#include <core/CHIPCore.h>
+#include <core/CHIPTLVDebug.hpp>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeMgr.h>
+#include <messaging/Flags.h>
+#include <protocols/Protocols.h>
+#include <support/CodeUtils.h>
+#include <support/DLLUtil.h>
+#include <support/logging/CHIPLogging.h>
+#include <system/SystemPacketBuffer.h>
+
+namespace chip {
+namespace app {
+/* A structure representing a path parameters to a event */
+struct EventPathParams
+{
+    EventPathParams() {}
+    EventPathParams(chip::EndpointId aEndpointId, chip::ClusterId aClusterId) : mEndpointId(aEndpointId), mClusterId(aClusterId) {}
+    chip::NodeId mNodeId;
+    chip::EndpointId mEndpointId;
+    chip::ClusterId mClusterId;
+    chip::EventId mEventId;
+};
+
+class ReadClient : public Messaging::ExchangeDelegate
+{
+public:
+    enum ClientState
+    {
+        kState_Uninitialized = 0,  ///< The client has not been initialized
+        kState_Initialized,        ///< The client has been initialized and is ready
+        kState_ReadRequestSending, ///< The client has sent out the read request message
+    };
+
+    enum EventID
+    {
+        kEvent_OnReadRequestPrepareNeeded = 1,
+
+        /**
+         * Generated on a receipt of a ReportDataRequest.
+         */
+        kEvent_OnReportDataRequest = 2,
+
+        /**
+         * Generated when InteractionModelEngine finishes processing a
+         * ReportDataRequest message.
+         */
+        kEvent_OnReportProcessed = 3,
+
+        /**
+         * Generated when the InteractionModelEngine encounters an event stream
+         * within the ReportDataRequest message.
+         */
+        kEvent_OnEventStreamReceived = 4,
+    };
+
+    // union of structures for each event some of them might be empty
+    union InEventParam
+    {
+        void Clear(void) { memset(this, 0, sizeof(*this)); }
+
+        struct
+        {
+            ReadClient * mpClient;
+        } mReadRequestPrepareNeeded;
+
+        struct
+        {
+            ReadClient * mpClient;
+            // Do not close the EC
+            Messaging::ExchangeContext * mpEC;
+
+            // Do not modify the message content
+            const PacketHeader * mpPacketHeader; ///< A pointer to the message information for the report data request
+        } mReportDataRequest;
+
+        struct
+        {
+            ReadClient * mpClient;
+        } mReportProcessed;
+
+        struct
+        {
+            chip::TLV::TLVReader * mpReader;
+            const PacketHeader * mpPacketHeader; ///< A pointer to the message information for the command request
```

Comment:
Comment is obvious and could be removed.

Should this have similar treatment as the packet header above (regarding 'do not change')? Or if not, maybe an explicit comment would be useful.

I am a bit concerned about complexity here if the rules differ by type.

Reply:
Done, thanks